refactoring.api/src/org/netbeans/modules/refactoring/api/RefactoringSession.java

```diff
@@ -246,27 +223,20 @@ private Problem reallyUndoRefactoring(boolean saveAfterDone) {
                     f.undoChange();
                 }
             }
-            for (Transaction commit : SPIAccessor.DEFAULT.getCommits(bag)) {
-                SPIAccessor.DEFAULT.check(commit, true);
-            }
-            UndoableWrapper wrapper = MimeLookup.getLookup("").lookup(UndoableWrapper.class);
+            SPIAccessor.DEFAULT.getCommits(bag).forEach(commit -> SPIAccessor.DEFAULT.check(commit, true));
+            UndoableWrapper undoableWrapper = MimeLookup.getLookup("").lookup(UndoableWrapper.class);
             for (ListIterator<Transaction> commitIterator = commits.listIterator(commits.size()); commitIterator.hasPrevious();) {
                 final Transaction commit = commitIterator.previous();
-                setWrappers(commit, wrapper);
+                undoableWrapper.setActive(true, this);
                 commit.rollback();
-                unsetWrappers(commit, wrapper);
-            }
-            wrapper.close();
-            for (Transaction commit : SPIAccessor.DEFAULT.getCommits(bag)) {
-                SPIAccessor.DEFAULT.sum(commit);
+                undoableWrapper.setActive(false, null);
             }
-
-            while (it.hasPrevious()) {
+            undoableWrapper.close();
```

Comment:
Should `undoableWrapper.close();` be within a finally block?

Reply:
Probably. Under the circumstance, that the code from within line 228 to 232 throws an unexpected exception. I'm not sure, if that is even possible. But it is a potential pitfall. I like your approach.